connection/connection.go

```diff
@@ -186,20 +195,25 @@ func connect(
 		lostConnection := false
 		reconnect := true
 
-		dialOptions = append(dialOptions, grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+		dialOptions = append(dialOptions, grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			logger := klog.FromContext(ctx)
 			if haveConnected && !lostConnection {
 				// We have detected a loss of connection for the first time. Decide what to do...
 				// Record this once. TODO (?): log at regular time intervals.
-				klog.Errorf("Lost connection to %s.", address)
+				logger.Error(nil, "Lost connection", "address", address)
 				// Inform caller and let it decide? Default is to reconnect.
 				if o.reconnect != nil {
-					reconnect = o.reconnect()
+					reconnect = o.reconnect(ctx)
 				}
 				lostConnection = true
 			}
 			if !reconnect {
 				return nil, errors.New("connection lost, reconnecting disabled")
 			}
+			timeout := 0 * time.Second
+			if deadline, ok := ctx.Deadline(); ok {
+				timeout = time.Until(deadline)
+			}
```

Comment:
I suspect that `net.DialTimeout(..., 0)` will not work. It might never occur, but perhaps switch between `net.Dial` and `net.DialTimeout` depending on whether there is a deadline?

Reply:
I think that using net.DialTimeout(..., 0) as mentioned above is not problematic, so I've not implemented a switch to net.Dial when timeout=0. I'd appreciate your thoughts on this.
(On a different note, but it seemed to me that using var timeout time.Duration is more readable than timeout := 0 * time.Second, so I have made that change as well.)